### DISCUSSION TITLE
Change `users.json` to YAML formatting

### DISCUSSION THREAD

**Discussion by redthirten** [OpenRCT2/OpenRCT2#16420] [Feature requests] [answered]
YAML formatting provides an easier, human readable experience, and less error prone editing of the `users` file, particularly for dedicated server admins who may not have direct access to the server via a client UI.

It would also provide an easier setup of the dedicated server's first human admin if they have file access, but not command line access.

**Comment by ZehMatt** [comment]
I disagree about YAML, I wouldn't mind something like TOML, json is also not particular bad I think. I think that the current format could be improved a bit.

**Reply by redthirten** [reply]
Why do you disagree about YAML? It has the least amount of special characters, making it one of the easiest object-oriented formats for a human to read and edit. And the `users` file contents are relatively non-critical and not lengthy, making any technical limitations of the format (ie. parsing speed, reliability, etc) kind of a moot point.

I'd settle for TOML; it looks like a nice "simplistic" version of JSON. But it only released last year and is kind of an obscure, specialized, and not a widely known/accepted format yet, making it a wild card in my opinion.

> I think that the current format could be improved a bit.

How can you improve the existing JSON format if its already following the format's specifications? Or do you simply mean you agree that it should be changed from JSON to anything else?

**Reply by ZehMatt** [reply]
I was talking about things like "groupId", would be alot easier to just use the literal group name. Regarding to why not YAML? I dislike it, its so easy to make mistakes, one space too much and its already screaming, not a very "user friendly" thing to me. Also TOML is been around since 2013 so its not exactly "obscure". Also TOML describes all the reasons on the website to why things like YAML are shit, that is actually why TOML exists.

**Reply by redthirten** [reply]
Good points. I just saw that TOML's 1.0 release was 2021 so I didn't know how "accepted" that was.

**Comment by duncanspumpkin** [comment, marked as answer]
Its a lose lose situation whatever is used.

**Reply by redthirten** [reply]
How so?...

**Reply by duncanspumpkin** [reply]
I think quite possibly every config format that has been chosen for both openrct2 and openloco people have suggested that it is the wrong one. Be that YAML, JSON, INI, ...

**Reply by redthirten** [reply]
True... but on the same token, I think a lot of people make these suggestions based on personal preference. I personally sway towards disliking JSON, but I still think it's pretty robust and should remain as the format for all internal OpenRCT2 configurations based on it's objective benefits. But I also make this suggustion for the `users` file based on objective benefits of YAML for a more user-facing file; not based on a subjective preference of YAML in general.

Then again... Having a project stick to one standard across the whole project would be *my* subjective opinion, but we got a mix already in this project with `config.ini` being INI, which I'm guessing was chosen over JSON due to it's readability (which is my original core premise). So maybe my suggustion should be to change it to INI to be in line with the project, but I don't think INI supports object-based data; does it?

**Reply by ZehMatt** [reply]
If anything we would use json for consistency reasons and not ini. Also I don't think there is a team member who actually thinks YAML is a good choice.

**Reply by redthirten** [reply]
> Also I don't think there is a team member who actually thinks YAML is a good choice.

My recommendation is based on this DM with @janisozaur. I personally think something custom and as simple-as-can-be should be used for anything potentially edited by humans (like my example in the DM).
![Capture](https://user-images.githubusercontent.com/4533989/149029161-361db1ed-642f-4214-8d62-7453f97ad25d.PNG)

**Reply by janisozaur** [reply]
Indeed, but I would accept better alternatives too. That said, I'm not sure if the change that would be this disruptive is justified. Before progressing with any of this, talking to some server hosters may be necessary.

**Reply by ZehMatt** [reply]
To be fair, explaining a noob on how to format YAML seems like a bigger deal to me, but that aside I don't think there is an urge to change it anytime soon, people seem to get around with what's currently present just fine.

**Reply by redthirten** [reply]
I'm going to close this suggustion because I think I'm going to take a page out of Minecraft's book: one of the largest and most prevalent game servers on the market. They use JSON for their `ops.json` admin file, and I think I've changed my mind to think that JSON is probably the best option for this kind of file. However, ***I do think*** that the `config.ini` file should remain INI, because INI is very similar to Minecraft's `server.properties` file, and is very easy to set for single, non-object-oriented entries (along with not complaining too much about whitespace like YAML does, which I totally forgot about; sorry).

This discussion has been very informative to me and I appreciate all the detailed feedback. Ultimately, I think everything should remain as-is.

**Reply by IntelOrca** [reply]
@ZehMatt I like YAML and it is a good alternative to .INI. It doesn't work so well for files that are often programmatically edited. Really good for configuration files though. For OpenLoco, we use it for the config file.
